### PR TITLE
Implement the raqote backend

### PR DESCRIPTION
**Pull request by Eijebong**
Rebased version of #23601 with some more stuff implemented. This passes some (not many) wpt tests.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23936)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -28,3 +28,4 @@ opt-level = 3
 # Those are here to dedupe winapi since mio is still using winapi 0.2.
 mio = { git = "https://github.com/servo/mio.git", branch = "servo" }
 iovec = { git = "https://github.com/servo/iovec.git", branch = "servo" }
+
```
Comment (Eijebong) [inline Cargo.toml]: :scissors:

File: components/canvas/raqote_backend.rs [hunk 2]
```diff
@@ -72,47 +80,90 @@ impl Backend for RaqoteBackend {
...
     pub fn set_miter_limit(&mut self, _val: f32) {
-        unimplemented!()
+        match self {
+            StrokeOptions::Raqote(options, _) => options.miter_limit = _val,
+        }
     }
     pub fn set_line_join(&mut self, _val: LineJoinStyle) {
-        unimplemented!()
+        match self {
+            StrokeOptions::Raqote(options, _) => options.join = _val.to_raqote_style(),
```
Comment (Eijebong) [inline components/canvas/raqote_backend.rs]: Rename to val

File: components/canvas/raqote_backend.rs [hunk 3]
```diff
@@ -72,47 +80,90 @@ impl Backend for RaqoteBackend {
...
+        }
     }
     pub fn set_line_cap(&mut self, _val: LineCapStyle) {
-        unimplemented!()
+        match self {
+            StrokeOptions::Raqote(options, _) => options.cap = _val.to_raqote_style(),
```
Comment (Eijebong) [inline components/canvas/raqote_backend.rs]: Rename to val too

File: components/canvas/raqote_backend.rs [hunk 4]
```diff
@@ -72,47 +80,90 @@ impl Backend for RaqoteBackend {
...
+            StrokeOptions::Raqote(options, _) => options,
+        }
     }
 }
 
 impl DrawOptions {
     pub fn set_alpha(&mut self, _val: f32) {
         match self {
-            DrawOptions::Raqote(()) => unimplemented!(),
+            DrawOptions::Raqote(draw_options) => draw_options.alpha = _val,
```
Comment (Eijebong) [inline components/canvas/raqote_backend.rs]: val here too

File: components/canvas/raqote_backend.rs [hunk 5]
```diff
@@ -194,69 +247,242 @@ impl GenericDrawTarget for raqote::DrawTarget {
...
+            rect.origin.x,
+            rect.origin.y,
+            rect.size.width,
+            rect.size.height,
+        );
+        raqote::DrawTarget::fill(
+            self,
+            &pb.finish(),
+            pattern.as_raqote(),
+            &raqote::DrawOptions::new(),
```
Comment (Eijebong) [inline components/canvas/raqote_backend.rs]: I think I could pass `draw_options.as_raqote()` here

File: components/canvas/raqote_backend.rs [hunk 6]
```diff
@@ -194,69 +247,242 @@ impl GenericDrawTarget for raqote::DrawTarget {
...
+        match self {
+            LineCapStyle::Butt => raqote::LineCap::Butt,
+            LineCapStyle::Round => raqote::LineCap::Round,
+            LineCapStyle::Square => raqote::LineCap::Square,
+        }
+    }
+}
+
+// TODO(pylbrecht)
+#[cfg(feature = "raqote_backend")]
```
Comment (Eijebong) [inline components/canvas/raqote_backend.rs]: :scissors: This is not needed anymore

**Comment by jdm** [comment]
Some new compilation failures when building with the azure backend:
```
[0m[1m[38;5;9merror[E0726][0m[0m[1m: implicit elided lifetime not allowed here[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mcomponents/canvas/azure_backend.rs:472:6[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m472[0m[0m [0m[0m[1m[38;5;12m| [0m[0mimpl Pattern {[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m     [0m[0m[1m[38;5;9m^^^^^^^[0m[0m[1m[38;5;12m-[0m[0m [0m[0m[1m[38;5;12mhelp: indicate the anonymous lifetime: `<'_>`[0m

[0m[1m[38;5;9merror[E0726][0m[0m[1m: implicit elided lifetime not allowed here[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mcomponents/canvas/azure_backend.rs:726:6[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m726[0m[0m [0m[0m[1m[38;5;12m| [0m[0mimpl Pattern {[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m     [0m[0m[1m[38;5;9m^^^^^^^[0m[0m[1m[38;5;12m-[0m[0m [0m[0m[1m[38;5;12mhelp: indicate the anonymous lifetime: `<'_>`[0m

[0m[1m[38;5;9merror[E0392][0m[0m[1m: parameter `'a` is never used[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mcomponents/canvas/canvas_data.rs:355:18[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m355[0m[0m [0m[0m[1m[38;5;12m| [0m[0mpub enum Pattern<'a> {[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m                 [0m[0m[1m[38;5;9m^^[0m[0m [0m[0m[1m[38;5;9munused parameter[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m    [0m[0m[1m[38;5;12m= [0m[0m[1mhelp[0m[0m: consider removing `'a` or using a marker such as `std::marker::PhantomData`[0m

[0m[1m[38;5;9merror[0m[0m[1m: aborting due to 3 previous errors[0m
```

**Comment by Eijebong** [comment]
It's now sitting there:
```
Ran 830 tests finished in 421.0 seconds.
  • 516 ran as expected. 13 tests skipped.
  • 195 tests crashed unexpectedly
  • 1 tests failed unexpectedly
  • 57 tests timed out unexpectedly
  • 61 tests had unexpected subtest results
```

The azure backend fix needs https://github.com/servo/rust-azure/pull/302 because Path needs to be cloned but is otherwise complete (ran all tests with the patch and they passed)

**Comment by Eijebong** [comment]
Now at:
```
Ran 830 tests finished in 296.0 seconds.
  • 568 ran as expected. 13 tests skipped.
  • 172 tests crashed unexpectedly
  • 1 tests failed unexpectedly
  • 16 tests timed out unexpectedly
  • 73 tests had unexpected subtest results
```

**Comment by jdm** [comment]
I'd love to merge these changes. Would it take much to address the various nits and rebase this PR?

**Review by jdm — APPROVED** [review]

**Comment by Eijebong** [comment]
@jdm there

**Comment by jdm** [comment]
@bors-servo r+
Thanks!

**Comment by Eijebong** [comment]
> [taskcluster:error] Task aborted - max run time exceeded

**Comment by jdm** [comment]
@bors-servo retry
* killed the rogue test processes on mac5

**Comment by Eijebong** [comment]
```
{
    "status": "CRASH", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/referrer-policy/generic/iframe-src-change.html", 
    "line": 192884, 
    "action": "test_result", 
    "expected": "OK"
}
```

**Pull request merged**